### PR TITLE
Save/Restore Temperature only in Thermal Runs

### PR DESCRIPTION
**Pull request by bska**
This commit makes the output and reload of temperature data contingent on the simulation run actually enabling thermal mode. This, in turn, increases the compatibility of Flow's restart mechanism with that of the commercial simulator's restart files.

### CODE REVIEW

**Comment by bska** [comment]
Note: This pull request depends on OPM/opm-common#546 and must not be merged before that one is installed.

**Comment by bska** [comment]
jenkins build this opm-common=546 please

File: ebos/ecloutputblackoilmodule.hh [hunk 1]
```diff
@@ -229,7 +229,9 @@ public:
         }
         // and oil pressure
         oilPressure_.resize(bufferSize, 0.0);
-        temperature_.resize(bufferSize, 0.0);
+
+        if (simulator_.vanguard().eclState().getSimulationConfig().isThermal())
```
Comment (andlaus) [inline ebos/ecloutputblackoilmodule.hh]: ebos is more generic than eclipse here: even if energy is not conserved, the temperature does not necessarily need to be uniform throughout the domain. granted, not much is influenced by temperature, but anyway...

(also, it is probably better to use `if (enableEnergy)` on the simulator side.)

File: ebos/eclwriter.hh [hunk 2]
```diff
@@ -318,13 +318,18 @@ public:
 
     void restartBegin()
     {
+        const auto& es = simulator_.vanguard().eclState();
```
Comment (andlaus) [inline ebos/eclwriter.hh]: s/es/eclState/

File: ebos/eclwriter.hh [hunk 3]
```diff
@@ -318,13 +318,18 @@ public:
...
+
         bool enableHysteresis = simulator_.problem().materialLawManager()->enableHysteresis();
-        bool enableSwatinit = simulator_.vanguard().eclState().get3DProperties().hasDeckDoubleGridProperty("SWATINIT");
+        bool enableSwatinit = es.get3DProperties().hasDeckDoubleGridProperty("SWATINIT");
+
+        bool enableTemperature = es.getSimulationConfig().isThermal();
```
Comment (andlaus) [inline ebos/eclwriter.hh]: `enableEnergy` should already be present? (also be aware that non-uniform temperature does not necessarily imply energy conservation.)

File: ebos/eclwriter.hh [hunk 4]
```diff
@@ -318,13 +318,18 @@ public:
...
+
         std::vector<Opm::RestartKey> solutionKeys{
             {"PRESSURE" , Opm::UnitSystem::measure::pressure},
             {"SWAT" ,     Opm::UnitSystem::measure::identity, static_cast<bool>(FluidSystem::phaseIsActive(FluidSystem::waterPhaseIdx))},
             {"SGAS" ,     Opm::UnitSystem::measure::identity, static_cast<bool>(FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx))},
-            {"TEMP" ,     Opm::UnitSystem::measure::temperature}, // always required for now
+            {"TEMP" ,     Opm::UnitSystem::measure::temperature, enableTemperature}, // restore if thermal run
```
Comment (andlaus) [inline ebos/eclwriter.hh]: adding a third argument to `Opm::RestartKey` for enabling/disabling it is a quite strange design decision IMO. the key should simply not be added if it does not make sense. maybe there were deeper reasons?

**Comment by andlaus** [comment]
@bska: are you going to fix this up, or shall it be closed?

**Comment by bska** [comment]
> are you going to fix this up, or shall it be closed?

There's nothing really to fix.  This, or something very similar to this, is necessary to enable restarting Flow from an ECLIPSE-compatible restart file.  In particular, we must not require that `TEMP` is present in the result set from which we're restarting unless we're actually in a thermal run.  That also suggests that we should not bother outputting a temperature vector to the result set unless we're in a thermal run.

Since these changes are in the files that directly implement ECLIPSE-style input/output, EBOS's notion of energy conservation is not really relevant.

File: ebos/eclwriter.hh (hunk 4 above)
Comment (bska) [reply ebos/eclwriter.hh]: > adding a third argument to `Opm::RestartKey` for enabling/disabling it is a quite strange design decision IMO.

That's not really what the flag does.  Rather it indicates whether or not the associated vector must be present in the result set when restarting.  If a requisite vector is not present, then the restart procedure (and simulation run) terminates.

**Comment by andlaus** [comment]
> There's nothing really to fix. This, or something very similar to this, is necessary to enable restarting Flow from an ECLIPSE-compatible restart file.

I did not mean anything fundamental, just responding to the review.

> That also suggests that we should not bother outputting a temperature vector to the result set unless we're in a thermal run.

I'm pretty sure all the `if` messiness that comes with this is not worth the benefits: storing temperature does not require a significant amount of memory and it also does not adversely affect performance (?, haven't checked but would be very surprised)

> Since these changes are in the files that directly implement ECLIPSE-style input/output, EBOS's notion of energy conservation is not really relevant.

since this is ebos core code, it is very well relevant for it ;). use `enableEnergy` instead of getting this directly from the parser to avoid inconsistencies within the simulator. if you really want to do it the Right Way (TM), you can change the flow main function to use the internalized parser flag to decide if energy ought to be enabled or not.

**Comment by bska** [comment]
> > That also suggests that we should not bother outputting a temperature vector to the result set unless we're in a thermal run.
> 
> I'm pretty sure all the `if` messiness that comes with this is not worth the benefits: storing temperature does not require a significant amount of memory and it also does not adversely affect performance (?, haven't checked but would be very surprised)

I suppose there's one piece of information that's not clearly communicated here.  If we want to be strictly compatible with ECLIPSE, then we're not **allowed** to store temperature data in the restart file unless it's a thermal run (i.e., unless either of the `TEMP` or `THERMAL` keywords are specified in `RUNSPEC`).  Conversely, we can't expect there to be temperature data in the restart file when restarting a non-thermal run from a strictly ECLIPSE-compatible restart file.

In other words, the appropriate condition for saving/restoring temperature data is something along the lines of
```C++
bool saveRestoreTemp = (is_strictly_eclipse_compatible_restart)
    ? (runspec.has("TEMP") || runspec.has("THERMAL"))
    : GET_PROP_VALUE(TypeTag, EnableEnergy);
```
I'll update the PR accordingly.

**Comment by andlaus** [comment]
> If we want to be strictly compatible with ECLIPSE, then we're not allowed to store temperature data in the restart file unless it's a thermal run 

I got this, and I agree that it shouldn't. my point is that we are guaranteed not to be in a thermal run if `enableEnergy` is false and we are guaranteed to be in one if it is true. IOW, the simulator does not currently care whether the parser flag is set or not.

**Comment by bska** [comment]
> my point is that we are guaranteed not to be in a thermal run if `enableEnergy` is false and we are guaranteed to be in one if it is true.

Okay, so that suggests that it's possible to have `enableEnergy == true` when the run does not specify either of the `TEMP` or `THERMAL` keywords.  That's fine if we're not trying to create ECLIPSE-compatible restart files.  It is not fine if we are trying to create ECLIPSE-compatible restart files.  ECLIPSE-compatible restart files must not have temperature data unless the run specifies `TEMP` or `THERMAL`.

**Comment by andlaus** [comment]
the simultor enables energy conservation if and only if the `EnableEnergy` property is set to `true`. it does not actually care about the parser here, but hopefully the property is set iff your new `EclipseState` flag is set to `true`.

(See it this way: starting an isothermal simulation for a deck that contains `TEMP` *might* make sense in some scenarios. At least I have much less trouble coming up with applications for this than for being able to set the `SummaryConfig` and `Schedule` objects externally.)

tl;dr: use `enableEnergy` within the simulator and make sure that it corresponds to the parser flag in `flow.cpp`!

**Comment by andlaus** [comment]
I'd like to merge this, but you still need to do the `simulator_.vanguard().eclState().getSimulationConfig().isThermal())` to `enableEnergy` rename.

File: ebos/eclwriter.hh (hunk 4 above)
Comment (andlaus) [reply ebos/eclwriter.hh]: I suppose that's the problem with flag arguments: its often not clear what they to at the call-site.

**Comment by bska** [comment]
> you still need to do the [`isThermal()` to `enableEnergy`] rename

Okay, I'll try.  That's a much bigger change though.

**Comment by andlaus** [comment]
jenkins build this update_data please

**Comment by andlaus** [comment]
> That's a much bigger change though.

should be approximately two lines of code...

**Comment by bska** [comment]
> > That's a much bigger change though.
> 
> should be approximately two lines of code...

In this repository, yes.  However, I will also need to touch at least opm-simulators, so there's more effort to coordinate and test everything.

**Comment by andlaus** [comment]
(I wanted to say that the opm-simulator changes are not required from a technical point of view, bit are nice to have because they reduce code duplication between opm-simulators and opm-common.)

**Comment by bska** [comment]
I'm closing this.  It is strictly superseded by #433.
